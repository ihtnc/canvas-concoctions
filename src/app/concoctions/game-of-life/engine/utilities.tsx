import { type MatrixCoordinate, type MatrixValue, PeekDirection, peek } from "@/utilities/matrix-operations"
import { isCellUndefinedOrTechnicallyAlive } from "./cell-state-operations"
import { type CellValue, CellState } from "./types"
import { type HSL, type BorderRadii } from "@/utilities/drawing-operations"
import { deepCopy } from "@/utilities/misc-operations"
import ENGINE_DATA from './data'
import { getCycleProgress } from "./cycle-operations"

type GenerateCellsFunction = (value: MatrixValue<CellValue>, coordinate: MatrixCoordinate) => MatrixValue<CellValue>
export const generateCells: GenerateCellsFunction = (value, coordinate) => {
  const { row, col } = coordinate
  if (row >= value.length) { return value }
  if (col >= value[row].length) { return value }

  value[row][col] = deepCopy(ENGINE_DATA.AliveCell)
  return value
}

type CellStateMap = { state: CellState, cells: Array<MatrixCoordinate> }
type GetStateMapFunction = (value: MatrixValue<CellValue>) => Array<CellStateMap>
export const getStateMap: GetStateMapFunction = (value): Array<CellStateMap> => {
  const sorted: Array<CellStateMap> = []

  for (let i = 0; i < value.length; i++) {
    for (let j = 0; j < value[i].length; j++) {
      const state = value[i][j].state
      let group = sorted.find(item => item.state === state)
      if (group === undefined) {
        const newLength = sorted.push({ state, cells: [] })
        group = sorted[newLength - 1]
      }
      group.cells.push({ row: i, col: j })
    }
  }

  return sorted
}

type CellDisplayMap = {
  state: CellState,
  color: HSL,
  opacity: number,
  width: number, height:
  number, cells: Array<MatrixCoordinate>
}
type GetDisplayMapFunction = (stateMap: Array<CellStateMap>, cycleIndex: number, width: number, height: number, aliveColor: HSL, dyingColor: HSL, growingColor: HSL) => Array<CellDisplayMap>
export const getDisplayMap: GetDisplayMapFunction = (stateMap, cycleIndex, width: number, height: number, aliveColor, dyingColor, growingColor): Array<CellDisplayMap> => {
  const sorted: Array<CellDisplayMap> = []
  const cycleProgress = getCycleProgress(cycleIndex)

  for (let i = 0; i < stateMap.length; i++) {
    if (stateMap[i].state === CellState.Dead) { continue }

    let color: HSL
    let cellWidth: number
    let cellHeight: number
    let opacity: number
    const state = stateMap[i].state
    switch (state) {
      case CellState.Dying:
        const dyingProgress = 1 - cycleProgress
        color = deepCopy(dyingColor)
        color.s = color.s * dyingProgress
        opacity = dyingProgress
        cellWidth = width * dyingProgress
        cellHeight = height * dyingProgress
        break
      case CellState.Growing:
        const growingProgress = cycleProgress
        color = deepCopy(growingColor)
        color.s = color.s * growingProgress
        opacity = growingProgress
        cellWidth = width * growingProgress
        cellHeight = height * growingProgress
        break
      default:
        color = deepCopy(aliveColor)
        opacity = 1
        cellWidth = width
        cellHeight = width
        break
    }

    const group: CellDisplayMap = {
      state,
      color,
      opacity,
      width: cellWidth,
      height: cellHeight,
      cells: []
    }
    for (let j = 0; j < stateMap[i].cells.length; j++) {
      group.cells.push(stateMap[i].cells[j])
    }

    sorted.push(group)
  }

  return sorted
}

export const getCellShape: (map: MatrixValue<CellValue>, coordinate: MatrixCoordinate, cycleIndex: number) => BorderRadii = (map, coordinate, cycleIndex) => {
  const { row, col } = coordinate
  const shape = deepCopy(ENGINE_DATA.CellShape)

  const cellTopLeft = peek(map, { row, col }, PeekDirection.UpperLeft)
  if (isCellUndefinedOrTechnicallyAlive(cellTopLeft)) {
    shape.tl = 0
  }

  const cellAbove = peek(map, { row, col }, PeekDirection.Up)
  if (isCellUndefinedOrTechnicallyAlive(cellAbove)) {
    shape.tl = 0
    shape.tr = 0
  }

  const cellTopRight = peek(map, { row, col }, PeekDirection.UpperRight)
  if (isCellUndefinedOrTechnicallyAlive(cellTopRight)) {
    shape.tr = 0
  }

  const cellLeft = peek(map, { row, col }, PeekDirection.Left)
  if (isCellUndefinedOrTechnicallyAlive(cellLeft)) {
    shape.tl = 0
    shape.bl = 0
  }

  const cellRight = peek(map, { row, col }, PeekDirection.Right)
  if (isCellUndefinedOrTechnicallyAlive(cellRight)) {
    shape.tr = 0
    shape.br = 0
  }

  const cellBottomLeft = peek(map, { row, col }, PeekDirection.LowerLeft)
  if (isCellUndefinedOrTechnicallyAlive(cellBottomLeft)) {
    shape.bl = 0
  }

  const cellBelow = peek(map, { row, col }, PeekDirection.Down)
  if (isCellUndefinedOrTechnicallyAlive(cellBelow)) {
    shape.bl = 0
    shape.br = 0
  }

  const cellBottomRight = peek(map, { row, col }, PeekDirection.LowerRight)
  if (isCellUndefinedOrTechnicallyAlive(cellBottomRight)) {
    shape.br = 0
  }

  return shape
}