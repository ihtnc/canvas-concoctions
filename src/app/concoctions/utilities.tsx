import { NavigationDetails as sandSimNav } from './sand-sim';
import { NavigationDetails as tagVisualiserNav } from './tag-visualiser';
import { NavigationDetails as gameOfLifeNav } from './game-of-life';

export type ConcoctionNavigation = {
  linkTitle: string,
  linkUrl: string,
  title: string,
  previewUrl?: string
};

type GetConcoctionsResponse = Array<ConcoctionNavigation>;
export const getConcoctions = (): GetConcoctionsResponse => {
  return [
    sandSimNav,
    tagVisualiserNav,
    gameOfLifeNav
  ];
};
