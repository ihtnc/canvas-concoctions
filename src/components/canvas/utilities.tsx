import {
  RenderLocation,
  type DebugLayerDrawHandler,
  type DebugLayerOptions,
  type DebugLayerRendererValue,
  type Use2DRenderLoopOptions,
  type Coordinates,
  type DebugLayerValue,
} from "./types";

export const DEFAULT_DEBUG_LAYER_OPTIONS: DebugLayerOptions = {
  location: RenderLocation.TopLeft,
  renderFps: true,
  renderSize: true,
  renderClientSize: true,
  renderPixelRatio: true
};

export const DEFAULT_OPTIONS: Use2DRenderLoopOptions = {
  clearEachFrame: true,
  debugLayerRenderer: false
};

export const getDebugLayerRenderer: (value?: DebugLayerRendererValue) => DebugLayerDrawHandler | null = (value) => {
  let options: DebugLayerOptions;
  options = DEFAULT_DEBUG_LAYER_OPTIONS;

  if (value === undefined || value === false) {
    return null;
  }

  if (typeof(value) === "number") {
    options.location = value as RenderLocation;
  }

  const { x, y } = value as Coordinates;
  if (x !== undefined && y !== undefined) {
    options.location = { x, y };
  }

  if (typeof(value) === "object") {
    const {
      location,
      renderFps,
      renderSize,
      renderClientSize,
      renderPixelRatio
    } = value as DebugLayerOptions;

    if (location !== undefined) { options.location = location; }
    if (renderFps !== undefined) { options.renderFps = renderFps; }
    if (renderSize !== undefined) { options.renderSize = renderSize; }
    if (renderClientSize !== undefined) { options.renderClientSize = renderClientSize; }
    if (renderPixelRatio !== undefined) { options.renderPixelRatio = renderPixelRatio; }
  }

  let renderer: DebugLayerDrawHandler;
  if(typeof value === "function") {
    renderer = value;
  }

  const getCoordinates: (value: string, debugValue: DebugLayerValue, context: CanvasRenderingContext2D) => Coordinates = (value, debugValue, context) => {
    let { x, y } = options.location as Coordinates;
    if (x !== undefined && y !== undefined) { return {x, y}; }

    const { width, actualBoundingBoxAscent, actualBoundingBoxDescent } = context.measureText(value);
    const textWidth = width;
    const textHeight = actualBoundingBoxAscent + actualBoundingBoxDescent;
    const offSet = 10;
    const leftX = 0 + offSet;
    const topY = 0 + offSet + offSet;
    const midX = (debugValue.width / 2) - (textWidth / 2);
    const midY = (debugValue.height / 2) - (textHeight / 2);
    const rightX = debugValue.width - textWidth - offSet;
    const bottomY = debugValue.height - textHeight;
    switch(options.location) {
      case RenderLocation.TopLeft:
        x = leftX;
        y = topY;
        break;
      case RenderLocation.TopCenter:
        x = midX;
        y = topY;
        break;
      case RenderLocation.TopRight:
        x = rightX;
        y = topY;
        break;
      case RenderLocation.MiddleLeft:
        x = leftX;
        y = midY;
        break;
      case RenderLocation.Center:
        x = midX;
        y = midY;
        break;
      case RenderLocation.MiddleRight:
        x = rightX;
        y = midY;
        break;
      case RenderLocation.BottomLeft:
        x = leftX;
        y = bottomY;
        break;
      case RenderLocation.BottomCenter:
        x = midX;
        y = bottomY;
        break;
      case RenderLocation.BottomRight:
        x = rightX;
        y = bottomY;
        break;
    }
    return { x, y };
  }

  renderer = (value, context) => {
    const fpsText = options.renderFps ? `fps: ${value.fps}; ` : '';
    const sizeText = options.renderSize ? `size: ${value.width}x${value.height}; ` : '';
    const clientText = options.renderClientSize ? `client: ${value.clientWidth}x${value.clientHeight}; ` : '';
    const ratioText = options.renderPixelRatio ? `ratio: ${value.pixelRatio}; ` : '';
    const debugText = `${fpsText}${sizeText}${clientText}${ratioText}`.trim();
    const { x, y } = getCoordinates(debugText, value, context);
    const originalFillStyle = context.fillStyle;
    context.fillStyle = "#000000";
    context.fillText(debugText, x, y);
    context.fillStyle = originalFillStyle;
  };

  return renderer;
}
