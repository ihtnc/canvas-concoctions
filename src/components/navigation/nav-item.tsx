import { useFloating, useHover, useInteractions } from '@floating-ui/react'
import Image from 'next/image'
import Link from 'next/link'
import { useState } from 'react'

type NavItemProps = {
  href: string,
  className?: string,
  title: string,
  previewHref?: string,
  isActive?: boolean
};

const NavItem = ({ href, className, title, previewHref, isActive }: NavItemProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const { refs, floatingStyles, context } = useFloating({
    open: isOpen,
    onOpenChange: setIsOpen
  })
  const hover = useHover(context)
  const { getReferenceProps, getFloatingProps } = useInteractions([hover])

  return (
    <>
      <Link
        ref={refs.setReference}
        href={href}
        className={`link ${className ?? ''} ${isActive ? 'active': ''}`}
        {...getReferenceProps()}
      >
        {title}
      </Link>
      {!isActive && previewHref && isOpen && (
        <Image
          className='w-72 h-auto border border-black'
          src={previewHref} alt={title}
          width='0' height='0'
          ref={refs.setFloating}
          style={floatingStyles}
          { ...getFloatingProps()}
        />
      )}
    </>
  )
}

export default NavItem