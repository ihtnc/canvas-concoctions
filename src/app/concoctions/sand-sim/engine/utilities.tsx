import { type ParticleColorMap, type ParticleValue } from "./types"
import ENGINE_DATA from './data'
import { type MatrixCoordinate, type MatrixValue, PeekDirection, peek } from "@/utilities/matrix-operations"
import { chooseRandom, deepCopy } from "@/utilities/misc-operations"
import { areHSLsEqual } from "@/utilities/drawing-operations"

export const isValueUndefinedOrNotEmpty = (particle?: ParticleValue): boolean => {
  return particle === undefined || !isValueEmpty(particle)
}

export const isValueEmpty = (particle: ParticleValue): boolean => {
  return particle.color === undefined || particle.value === ENGINE_DATA.EmptyParticle.value
}

export const calculateDistance = (time: number): number => {
  // distance = (initialVelocity * time) + (0.5 * acceleration * time^2)
  const initial = time * ENGINE_DATA.InitialVelocity
  const accelaration = (0.5 * ENGINE_DATA.Acceleration * time * time)
  return initial + accelaration
}

export const getNextCoordinate = (value: MatrixValue<ParticleValue>, current: MatrixCoordinate): MatrixCoordinate | undefined => {
  let direction: PeekDirection = PeekDirection.Down
  let retryCount: number = 0

  while (retryCount < 2) {
    const particleBelow = peek(value, current, direction)
    if (particleBelow === undefined) { return undefined }

    if (isValueEmpty(particleBelow)) {
      const next: MatrixCoordinate = {
        row: current.row + 1,
        col: current.col
      }

      if (direction === PeekDirection.LowerLeft) { next.col = current.col - 1 }
      else if (direction === PeekDirection.LowerRight) { next.col = current.col + 1 }

      return next
    }

    const option = chooseRandom(1, 100)
    if (option <= 48) { direction = PeekDirection.LowerLeft }
    else if (option <= 52) { return undefined }
    else { direction = PeekDirection.LowerRight }

    retryCount++
  }

  return undefined
}

export const getNextValue = (particle: ParticleValue): ParticleValue => {
  const nextValue = deepCopy(particle)
  nextValue.value = nextValue.value + ENGINE_DATA.IncrementValue
  return deepCopy(nextValue)
}

export const getColorMap = (value: MatrixValue<ParticleValue>): Array<ParticleColorMap> => {
  const sorted: Array<ParticleColorMap> = []

  for (let i = 0; i < value.length; i++) {
    for (let j = 0; j < value[i].length; j++) {
      if (value[i][j].color === undefined) { continue }

      const color = value[i][j].color!
      let group = sorted.find(item => areHSLsEqual(item.color, color))
      if (group === undefined) {
        const newLength = sorted.push({ color: deepCopy(color), particles: [] })
        group = sorted[newLength - 1]
      }
      group.particles.push({ row: i, col: j })
    }
  }

  return sorted
}