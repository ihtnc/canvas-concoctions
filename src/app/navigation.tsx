'use client';

import { getConcoctions } from './concoctions/utilities';
import Link from 'next/link';
import { usePathname } from 'next/navigation';

const Navigation = () => {
  const links = getConcoctions();
  const pathName = usePathname();

  const activeLink = links.find(l => `/concoctions/${l.linkUrl}` === pathName);
  const isHomeActive = pathName === '/';

  return (
    <nav>
      <ul className='flex flex-row gap-2'>
        <li key='/' className='flex'>
          <Link
            href='/'
            className={`link ${isHomeActive ? 'active' : ''}`}>
            Home
          </Link>
        </li>
        {links.map((value) => (
          <li key={value.linkUrl} className='flex'>
            <Link
              href={`/concoctions/${value.linkUrl}`}
              className={`link ${value.linkUrl === activeLink?.linkUrl ? 'active' : ''}`}>
              {value.linkTitle}
            </Link>
          </li>
        ))}
      </ul>
    </nav>
  );
};

export default Navigation;