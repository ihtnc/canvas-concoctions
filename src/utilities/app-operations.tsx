import { useSearchParams } from "next/navigation"

type AppDisplay = {
  nav: boolean,
  title: boolean
}

export const useAppDisplay = (): AppDisplay => {
  const searchParams = useSearchParams()

  return {
    nav: searchParams.has("no-nav") === false,
    title: searchParams.has("no-title") === false
  }
}