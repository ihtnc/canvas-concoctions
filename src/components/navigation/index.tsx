'use client'

import { usePathname } from 'next/navigation'
import { getConcoctions } from '@/app/concoctions/utilities'
import NavItem from './nav-item'
import { useAppDisplay } from '@/utilities/app-operations'

type NavigationProps = {
  baseUrl?: string
};

const Navigation = ({ baseUrl }: NavigationProps) => {
  const links = getConcoctions()
  const pathName = usePathname()
  const concoctionId = pathName.split('/').pop()
  const { nav: showNav } = useAppDisplay()
  const itemBaseUrl = baseUrl !== undefined ? `/${baseUrl.replace(/^\//, '').replace(/\/$/, '')}` : ''

  const constructPath = (linkUrl: string) => {
    const trimUrl = linkUrl.replace(/^\//, '')
    return `${itemBaseUrl}/${trimUrl}`
  }

  return showNav ? (
    <section className="flex flex-col self-center">
      <nav>
        <ul className='flex flex-row gap-2'>
          <li key='/' className='flex'>
            <NavItem
              href='/'
              title='Home'
              isActive={pathName === '/'} />
          </li>
          {links.map((value) => (
            <li key={value.linkUrl} className='flex'>
              <NavItem
                href={`${constructPath(value.linkUrl)}`}
                title={value.linkTitle}
                previewHref={value.previewUrl ? value.previewUrl : '/previews/default.gif'}
                isActive={value.linkUrl === concoctionId}
              />
            </li>
          ))}
        </ul>
      </nav>
    </section>
  ) : null
}

export default Navigation