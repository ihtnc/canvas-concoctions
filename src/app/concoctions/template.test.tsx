import { render } from '@testing-library/react'
import ConcoctionsTemplate from './template'
import { Mock, afterEach, beforeEach, describe, expect, test, vi } from 'vitest'
import { getConcoction } from './utilities'
import { useSelectedLayoutSegment } from 'next/navigation'
import { useAppDisplay } from '@/utilities/app-operations'

vi.mock('./utilities', () => ({
  getConcoction: vi.fn()
}))

vi.mock('next/navigation', () => ({
  useSelectedLayoutSegment: vi.fn()
}))

vi.mock('@/utilities/app-operations', () => ({
  useAppDisplay: vi.fn()
}))

describe('ConcoctionsTemplate component', () => {
  let getConcoctionMock: Mock
  let appDisplay = { title: true }
  let useAppDisplayMock: Mock
  let useSelectedLayoutSegmentMock: Mock

  beforeEach(() => {
    getConcoctionMock = getConcoction as Mock

    appDisplay = { title: true }
    useAppDisplayMock = useAppDisplay as Mock
    useAppDisplayMock.mockReturnValue(appDisplay)

    useSelectedLayoutSegmentMock = useSelectedLayoutSegment as Mock
  })

  afterEach(() => {
    vi.resetAllMocks()
  })

  test('should try to get title of current concoction', () => {
    const segment = 'test-concoction'
    useSelectedLayoutSegmentMock.mockReturnValue(segment)

    render(<ConcoctionsTemplate><></></ConcoctionsTemplate>)

    expect(useSelectedLayoutSegment).toHaveBeenCalled()
    expect(getConcoction).toHaveBeenCalledWith(segment)
  })

  test('should check for app display status', () => {
    render(<ConcoctionsTemplate><></></ConcoctionsTemplate>)

    expect(useAppDisplay).toHaveBeenCalled()
  })

  test('should render the title of the current concoction correctly', () => {
    getConcoctionMock.mockReturnValue({ title: 'Test Concoction' })

    const { container } = render(<ConcoctionsTemplate><></></ConcoctionsTemplate>)

    expect(container).toMatchSnapshot()
  })

  test('should render the title of the current concoction if it is not recognised', () => {
    getConcoctionMock.mockReturnValue(undefined)

    const { container } = render(<ConcoctionsTemplate><></></ConcoctionsTemplate>)

    expect(container).toMatchSnapshot()
  })

  test('should render the children correctly', () => {
    const { container } = render(<ConcoctionsTemplate><>Test Content</></ConcoctionsTemplate>)
    expect(container).toMatchSnapshot()
  })

  test('should not render title if title is hidden in app display', () => {
    appDisplay.title = false
    getConcoctionMock.mockReturnValue({ title: 'Test Concoction' })

    const { container } = render(<ConcoctionsTemplate><>Test Content</></ConcoctionsTemplate>)

    expect(container).toMatchSnapshot()
  })
})