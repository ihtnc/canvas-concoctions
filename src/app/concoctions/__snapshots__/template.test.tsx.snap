// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ConcoctionsTemplate component > should not render title if title is hidden in app display 1`] = `
<div>
  <section
    class="flex flex-col w-full h-full gap-2"
  >
    Test Content
  </section>
</div>
`;

exports[`ConcoctionsTemplate component > should render the children correctly 1`] = `
<div>
  <section
    class="flex flex-col w-full h-full gap-2"
  >
    <h1
      class="self-center"
    >
      Concoction
    </h1>
    Test Content
  </section>
</div>
`;

exports[`ConcoctionsTemplate component > should render the title of the current concoction correctly 1`] = `
<div>
  <section
    class="flex flex-col w-full h-full gap-2"
  >
    <h1
      class="self-center"
    >
      Test Concoction
    </h1>
  </section>
</div>
`;

exports[`ConcoctionsTemplate component > should render the title of the current concoction if it is not recognised 1`] = `
<div>
  <section
    class="flex flex-col w-full h-full gap-2"
  >
    <h1
      class="self-center"
    >
      Concoction
    </h1>
  </section>
</div>
`;
